pkgs/development/python-modules/pywebview/default.nix

```diff
@@ -55,16 +55,16 @@ buildPythonPackage rec {
     # QStandardPaths: XDG_RUNTIME_DIR not set
     export XDG_RUNTIME_DIR=$HOME/xdg-runtime-dir
 
-    pushd tests
-    substituteInPlace run.sh \
-      --replace "PYTHONPATH=.." "PYTHONPATH=$PYTHONPATH" \
-      --replace "pywebviewtest test_js_api.py::test_concurrent ''${PYTEST_OPTIONS}" "# skip flaky test_js_api.py::test_concurrent"
+    cat > tests/run.sh <<"END"
+    #!/usr/bin/env bash
+    export PYWEBVIEW_LOG=debug
+    python3 -m pytest --deselect tests/test_js_api.py::test_concurrent
+    END
```

Comment:
Nit: writeShellApplication would be a lot nicer here.

Reply:
@Atemu Hmmm.… I agree. The problem with that is that the file currently doesn't have an existing let binding where I could put that, which'd mean a huge and annoying to undo diff, or some hard to read Nix interpolation inlining (which.… isn't that much worse than a here-doc I guess), and after everything's said and done it'd *still* have to get passed through the Qt wrapper which'd then have to be changed to the one that isn't inplace.

Like, if it wasn't for the Qt wrapper I would've already changed this to something like `env PYWEBVIEW_LOG=debug python3 ...` and inlined the whole thing without a here-doc, but somehow every solution just feels equally awkward in a different way.

[insert me reading up on the entire makeWrapper infra here]

Okay, how about this?

```nix
  checkPhase = ''
    # Cannot create directory /homeless-shelter/.... Error: FILE_ERROR_ACCESS_DENIED
    export HOME=$TMPDIR
    # QStandardPaths: XDG_RUNTIME_DIR not set
    export XDG_RUNTIME_DIR=$HOME/xdg-runtime-dir

    # a Qt wrapper is required to run the Qt backend
    # since the upstream script does not have a way to disable tests individually pytest is used directly instead
    makeQtWrapper "$(command -v pytest)" tests/run.sh \
      --set PYWEBVIEW_LOG debug \
      --add-flags "--deselect tests/test_js_api.py::test_concurrent"

    xvfb-run -s '-screen 0 800x600x24' tests/run.sh
  '';
```

It is equally awkward but it does handle all of the things in one place.
I made use of the `command -v` idiom to get the path at runtime, this avoids edge-cases when overriding since nothing is interpolated into the script so *PATH* changes during build time still work as expected (and the wrapper infra requires a file rather than a command).

Still not happy with it, but it seems to be the least cursed IMHO.
Would that solve your nit without creating more nits?
If so I'll put it in the PR.

Reply:
No that looks fine actually. You could even add the other two env vars to the wrapper too rather than exporting them in the global script state.

Reply:
> and your shebang patching didn't make sense either now that I look at this closer

Someone noticed ^^"

> You could even add the other two env vars to the wrapper too rather than exporting them in the global script state.

I thought of that, but I'm not sure whether they're needed by the xvfb-run, and I kind of do want to keep the stack order (xvfb-run ⇒ Qt wrapper ⇒ tests) since that's how it works at runtime for real applications too usually.
Also the way it's written the test script can *technically* be used outside the build env, as soon as it has the HOME stuff set that is no longer true (not tragic by any means of course, but somehow my code-senses are tingling when I try to move that bit of code so I'm wary of doing so).
However given that this may include some Xauthority stuff and whatnot I suspect that setting a working *HOME* for xvfb-run is meaningful (though I haven't exactly tested it otherwise).
I do however dislike the exports in the phase since the phases are sourced and not subshelled, which means they could technically affect later phases (not that it matters much, this is just me nit-picking now).

That was a lot of parenthesis.
Anyway, what about this then?

```nix
  checkPhase = ''
    # a Qt wrapper is required to run the Qt backend
    # since the upstream script does not have a way to disable tests individually pytest is used directly instead
    makeQtWrapper "$(command -v pytest)" tests/run.sh \
      --set PYWEBVIEW_LOG debug \
      --add-flags "--deselect tests/test_js_api.py::test_concurrent"

    # HOME and XDG directories are required for the tests
    env \
      HOME=$TMPDIR \
      XDG_RUNTIME_DIR=$TMPDIR/xdg-runtime-dir \
      xvfb-run -s '-screen 0 800x600x24' tests/run.sh
  '';
```

The tests still pass (and I'm inching closer to moving into a forest without computers).

One of the biggest gripes I have with shell scripts is that I can't put comments in continuation lines, which means comments have to move up a little so I condensed them which I guess is fine (just wanted to explain why).
I also wanted to have a `--` line to separate command from vars, but *env* does not like that.

Reply:
I've committed it for now, and also rebased against master while I was at it.
Feel free to throw in more suggestions or nits, but at this point I feel like this is ready to be merged (it didn't feel like it was quite there before).

Either way, thank you very much for the feedback!